### PR TITLE
Failure to find intersects of points on corners/edges of irregular quad.

### LINKED ISSUE #100: Failure to find intersects of points on corners/edges of irregular quad.
The Quad2D class can sometimes fail to find the intersect of a point which lies on the corner or edge of an irregular quad. Draft [PR#101](https://github.com/ecmwf/atlas/pull/101) demonstrates the occurrence of this bug.

### PR DESCRIPTION
**Pull request by odlomax**
This is a draft PR associated with #100 

An irregular Quad2D object sometimes fails to evaluate an intersect lying on a corner or edge of the quad.

fixes #100

### CODE REVIEW

**Comment by odlomax** [comment]
@twsearle has given me a thumbs up on this. @wdeconinck are the changes okay with you?

File: src/atlas/grid/detail/partitioner/MatchingMeshPartitionerCubedSphere.cc [hunk 1]
```diff
@@ -19,14 +20,17 @@ void MatchingMeshPartitionerCubedSphere::partition(const Grid& grid, int partiti
     // Make cell finder from owned mesh cells.
     const auto finder = interpolation::method::cubedsphere::CellFinder(prePartitionedMesh_);
 
+    // Numeric tolerance should scale with N.
+    const auto N = CubedSphereGrid(prePartitionedMesh_.grid()).N();
+    const auto tolerance = 2. * std::numeric_limits<double>::epsilon() * N;
```
Comment (wdeconinck) [inline src/atlas/grid/detail/partitioner/MatchingMeshPartitionerCubedSphere.cc:25]:
```suggestion
    const auto epsilon     = 2. * std::numeric_limits<double>::epsilon() * N;
    const auto edgeEpsilon = epsilon;
    const size_t listSize  = 4;
```

File: src/atlas/grid/detail/partitioner/MatchingMeshPartitionerCubedSphere.cc [hunk 2]
```diff
@@ -19,14 +20,17 @@ void MatchingMeshPartitionerCubedSphere::partition(const Grid& grid, int partiti
...
+
     // Loop over grid and set partioning[].
     auto lonlatIt = grid.lonlat().begin();
     for (gidx_t i = 0; i < grid.size(); ++i) {
         // This is probably more expensive than it needs to be, as it performs
         // a dry run of the cubedsphere interpolation method.
         const auto& lonlat = *lonlatIt;
-        partitioning[i]    = finder.getCell(lonlat).isect ? mpi::rank() : -1;
-        ;
+        partitioning[i]    = finder.getCell(lonlat, 4, tolerance, tolerance).isect ? mpi::rank() : -1;
```
Comment (wdeconinck) [inline src/atlas/grid/detail/partitioner/MatchingMeshPartitionerCubedSphere.cc:33]:
```suggestion
        partitioning[i]    = finder.getCell(lonlat, listSize, edgeEpsilon, epsilon).isect ? mpi::rank() : -1;
```

File: src/atlas/interpolation/element/Quad2D.cc [hunk 3]
```diff
@@ -51,37 +51,55 @@ method::Intersect Quad2D::intersects(const PointXY& r, double edgeEpsilon, doubl
...
+
+                double A = -b * inv_two_a;
+                double B = sqrt_det * inv_two_a;
+
+                // Scale epsilon by magnitude of A or B.
+                double edgeEpsilonScaled = edgeEpsilon *
+                                           std::max({1., std::abs(A), std::abs(B)});
+
+                return check_weight(wght = A + B, edgeEpsilonScaled) ? true :
+                       check_weight(wght = A - B, edgeEpsilonScaled) ? true : false;
```
Comment (wdeconinck) [inline src/atlas/interpolation/element/Quad2D.cc:94]: It looks really elegant to have "wght = A + B" or "wght = A - B" inline, but it is a bit confusing. I'm sure you were thinking it too 😉.
Also, perhaps we could spell "wght" out as "weight" 😏

**Review by wdeconinck — CHANGES_REQUESTED** [review]
Looks great! No issues besides a bit of nitpicking.

**Review by wdeconinck — APPROVED** [review]

**Pull request merged**